File: Python/Without_SOLID.py
class Employee:
    def __init__(self, id: int, name: str, type: str, salary: float, email: str):
        self._id = id
        self._name = name
        self._type = type
        self._salary = salary
        self._email = email

    def get_id(self) -> int:
        return self._id

    def get_name(self) -> str:
        return self._name

    def get_type(self) -> str:
        return self._type

    def get_salary(self) -> float:
        return self._salary

    def get_email(self) -> str:
        return self._email


class EmployeeService:
    def process_employee(self, employee: Employee, notification_type: str) -> None:
        # =====================
        # Validation
        # =====================
        if not employee.get_name():
            raise RuntimeError("Invalid Name")
        if "@" not in employee.get_email():
            raise RuntimeError("Invalid Email")

        # =====================
        # Salary Calculation
        # =====================
        emp_type = employee.get_type()
        if emp_type == "FULL_TIME":
            final_salary = employee.get_salary()
        elif emp_type == "INTERN":
            final_salary = employee.get_salary() * 0.5
        elif emp_type == "CONTRACT":
            final_salary = employee.get_salary() * 1.2
        else:
            raise RuntimeError("Unknown Employee Type")

        print(f"Salary : {final_salary}")

        # =====================
        # Database
        # =====================
        print("Connecting to MySQL...")
        print("Saving Employee...")
        print("Employee Saved")

        # =====================
        # Notification
        # =====================
        if notification_type == "EMAIL":
            print(f"Sending Email to {employee.get_email()}")
        elif notification_type == "SMS":
            print("Sending SMS")
        elif notification_type == "WHATSAPP":
            print("Sending WhatsApp Message")

        # =====================
        # Audit
        # =====================
        print("Writing Audit Log")

        # =====================
        # Analytics
        # =====================
        print("Publishing Analytics Event")

        # =====================
        # Invoice
        # =====================
        print("Generating Salary Slip")

        # =====================
        # Cloud Upload
        # =====================
        print("Uploading Salary Slip to AWS S3")

        # =====================
        # Cache
        # =====================
        print("Updating Redis Cache")

        # =====================
        # Report
        # =====================
        print("Generating Monthly Report")

        print("Employee Processing Completed")


if __name__ == "__main__":
    service = EmployeeService()
    emp = Employee(101, "Pradeep", "FULL_TIME", 100000, "pradeep@gmail.com")
    service.process_employee(emp, "EMAIL")

File: C++/Without_SOLID.cpp
#include <iostream>
#include <string>
#include <stdexcept>

class Employee {
    int id;
    std::string name;
    std::string type;
    double salary;
    std::string email;

public:
    Employee(int id, const std::string& name, const std::string& type, double salary, const std::string& email)
        : id(id), name(name), type(type), salary(salary), email(email) {}

    int get_id() const { return id; }
    const std::string& get_name() const { return name; }
    const std::string& get_type() const { return type; }
    double get_salary() const { return salary; }
    const std::string& get_email() const { return email; }
};

class EmployeeService {
public:
    void process_employee(const Employee& employee, const std::string& notification_type) {
        // =====================
        // Validation
        // =====================
        if (employee.get_name().empty()) {
            throw std::runtime_error("Invalid Name");
        }
        if (employee.get_email().find('@') == std::string::npos) {
            throw std::runtime_error("Invalid Email");
        }

        // =====================
        // Salary Calculation
        // =====================
        double final_salary = 0;
        const std::string& emp_type = employee.get_type();
        if (emp_type == "FULL_TIME") {
            final_salary = employee.get_salary();
        } else if (emp_type == "INTERN") {
            final_salary = employee.get_salary() * 0.5;
        } else if (emp_type == "CONTRACT") {
            final_salary = employee.get_salary() * 1.2;
        } else {
            throw std::runtime_error("Unknown Employee Type");
        }

        std::cout << "Salary : " << final_salary << std::endl;

        // =====================
        // Database
        // =====================
        std::cout << "Connecting to MySQL..." << std::endl;
        std::cout << "Saving Employee..." << std::endl;
        std::cout << "Employee Saved" << std::endl;

        // =====================
        // Notification
        // =====================
        if (notification_type == "EMAIL") {
            std::cout << "Sending Email to " << employee.get_email() << std::endl;
        } else if (notification_type == "SMS") {
            std::cout << "Sending SMS" << std::endl;
        } else if (notification_type == "WHATSAPP") {
            std::cout << "Sending WhatsApp Message" << std::endl;
        }

        // =====================
        // Audit
        // =====================
        std::cout << "Writing Audit Log" << std::endl;

        // =====================
        // Analytics
        // =====================
        std::cout << "Publishing Analytics Event" << std::endl;

        // =====================
        // Invoice
        // =====================
        std::cout << "Generating Salary Slip" << std::endl;

        // =====================
        // Cloud Upload
        // =====================
        std::cout << "Uploading Salary Slip to AWS S3" << std::endl;

        // =====================
        // Cache
        // =====================
        std::cout << "Updating Redis Cache" << std::endl;

        // =====================
        // Report
        // =====================
        std::cout << "Generating Monthly Report" << std::endl;

        std::cout << "Employee Processing Completed" << std::endl;
    }
};

int main() {
    EmployeeService service;
    Employee emp(101, "Pradeep", "FULL_TIME", 100000, "pradeep@gmail.com");
    service.process_employee(emp, "EMAIL");
    return 0;
}
